File: android/app/src/main/java/com/aurea/aurea/ai/GuardaDaSessao.kt
package com.aurea.aurea.ai

import android.app.Application
import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.File

/**
 * A geração da IA gravada no aparelho.
 *
 * Existe por causa do Rewarded: abrir um anúncio traz a Activity para trás e,
 * num aparelho de verdade com pouca memória, o sistema chega a matar o
 * processo. Sem isto, o `prompt_id` ia embora com o processo e o H3 terminava
 * sozinho do outro lado — o usuário ficava em "Gerando..." para sempre.
 *
 * O que é gravado é o mínimo para RETOMAR a MESMA geração: o id, o prompt_id, o
 * endereço que a aceitou, e o que já foi conquistado (geração concluída,
 * recompensa). Nada de segredo: `prompt_id` e endereço não abrem nada sozinhos.
 *
 * Regra que não muda: retomar NUNCA manda outro `POST /prompt`. Ou a geração é
 * acompanhada pelo prompt_id que já existe, ou não é acompanhada — repetir o
 * POST geraria um segundo job e cobraria a A100 duas vezes pelo mesmo vídeo.
 */
class GuardaDaSessao(app: Application) {

    private val prefs = app.getSharedPreferences("aurea_ai_sessao", Context.MODE_PRIVATE)

    private val chave = "sessao"

    fun gravar(s: AiGenerationSession) {
        val o = JSONObject().apply {
            put("generationId", s.generationId)
            put("promptId", s.promptId)
            put("endpointUsado", s.endpointUsado)
            put("generationCompleted", s.generationCompleted)
            put("rewardEarned", s.rewardEarned)
            put("generationStarted", s.generationStarted)
            put("adClosedEarly", s.adClosedEarly)
            put("arquivo", s.result?.absolutePath)
            put("erro", s.erro)
            put("adError", s.adError)
            put("status", s.status.name)
            put("pedido", JSONObject().apply {
                put("modo", s.pedido.modo)
                put("prompt", s.pedido.prompt)
                put("promptNegativo", s.pedido.promptNegativo)
                put("duracao", s.pedido.duracao)
                put("aspecto", s.pedido.aspecto)
                put("resolucao", s.pedido.resolucao)
                put("fps", s.pedido.fps)
                put("audio", s.pedido.audio)
                put("semente", s.pedido.semente)
                put("turbo", s.pedido.turbo)
                put("assetId", s.pedido.assetId)
            })
        }
        prefs.edit().putString(chave, o.toString()).apply()
    }

    fun limpar() {
        prefs.edit().remove(chave).apply()
    }

    /** A sessão gravada, ou nulo quando não há nenhuma (ou a gravada está ilegível). */
    fun ler(): AiGenerationSession? {
        val texto = prefs.getString(chave, null) ?: return null
        return try {
            val o = JSONObject(texto)
            val p = o.getJSONObject("pedido")
            val caminho = o.optString("arquivo").ifBlank { null }
            val arquivo = caminho?.let { File(it) }
            val status = runCatching { SessaoStatus.valueOf(o.getString("status")) }
                .getOrDefault(SessaoStatus.Gerando)
            if (status == SessaoStatus.Liberado || status == SessaoStatus.Falhou) {
                // Já terminou da última vez: não há o que retomar.
                limpar()
                return null
            }
            AiGenerationSession(
                generationId = o.getString("generationId"),
                pedido = Pedido(
                    modo = p.getString("modo"),
                    prompt = p.getString("prompt"),
                    promptNegativo = p.optString("promptNegativo"),
                    duracao = p.optInt("duracao", 5),
                    aspecto = p.optString("aspecto", "16:9"),
                    resolucao = p.optString("resolucao", "standard"),
                    fps = p.optInt("fps", 24),
                    audio = p.optBoolean("audio", true),
                    semente = p.optInt("semente", -1),
                    turbo = p.optBoolean("turbo", true),
                    assetId = p.optString("assetId").ifBlank { null },
                ),
                promptId = o.optString("promptId").ifBlank { null },
                endpointUsado = o.optString("endpointUsado").ifBlank { null },
                generationCompleted = o.optBoolean("generationCompleted"),
                rewardEarned = o.optBoolean("rewardEarned"),
                generationStarted = o.optBoolean("generationStarted"),
                adClosedEarly = o.optBoolean("adClosedEarly"),
                result = if (arquivo?.isFile == true) arquivo else null,
                erro = o.optString("erro").ifBlank { null },
                adError = o.optString("adError").ifBlank { null },
                status = status,
            )
        } catch (e: Exception) {
            Log.w("AureaAI", "[AUREA AI] ERROR = sessão gravada ilegível: $e")
            limpar()
            null
        }
    }
}

File: engine/platform/ios/app/Ai/GuardaDaSessao.swift
// =============================================================================
//  Aurea iOS — a geração da IA gravada no aparelho.
//
//  Porte de android/.../ai/GuardaDaSessao.kt. Existe pelo mesmo motivo: o
//  Rewarded traz o app para trás e o sistema pode matar o processo. Sem isto, o
//  `prompt_id` ia embora com o processo e o H3 terminava sozinho do outro lado
//  — a tela ficava em "Gerando..." para sempre.
//
//  Regra que não muda: retomar NUNCA manda outro `POST /prompt`. Ou a geração é
//  acompanhada pelo prompt_id que já existe, ou não é acompanhada — repetir o
//  POST geraria um segundo job e cobraria a A100 duas vezes pelo mesmo vídeo.
// =============================================================================
import Foundation

final class GuardaDaSessao {

    static let shared = GuardaDaSessao()

    private let chave = "aurea_ai_sessao"
    private let defaults = UserDefaults.standard

    func write(_ s: AiGenerationSession) {
        var d: [String: Any] = [
            "generationId": s.generationId,
            "generationCompleted": s.generationCompleted,
            "rewardEarned": s.rewardEarned,
            "generationStarted": s.generationStarted,
            "adClosedEarly": s.adClosedEarly,
            "status": s.status.rawValue,
            "mode": s.request.mode,
            "prompt": s.request.prompt,
            "negativePrompt": s.request.negativePrompt,
            "duration": s.request.duration,
            "aspect": s.request.aspect,
            "resolution": s.request.resolution,
            "fps": s.request.fps,
            "audio": s.request.audio,
            "seed": s.request.seed,
            "turbo": s.request.turbo,
        ]
        if let v = s.promptId { d["promptId"] = v }
        if let v = s.endpointUsed { d["endpointUsed"] = v }
        if let v = s.error { d["error"] = v }
        if let v = s.adError { d["adError"] = v }
        if let v = s.result { d["file"] = v.path }
        if let v = s.request.imageRef { d["imageRef"] = v }
        defaults.set(d, forKey: chave)
    }

    func clear() { defaults.removeObject(forKey: chave) }

    /// A sessão gravada, ou nulo quando não há nenhuma (ou a gravada está ilegível).
    func read() -> AiGenerationSession? {
        guard let d = defaults.dictionary(forKey: chave) else { return nil }
        guard let id = d["generationId"] as? String else { clear(); return nil }

        let status = AiSessionStatus(rawValue: d["status"] as? String ?? "") ?? .generating
        // Já terminou da última vez: não há o que retomar.
        if status == .unlocked || status == .failed { clear(); return nil }

        var request = AiRequest(mode: d["mode"] as? String ?? "text_to_video",
                                prompt: d["prompt"] as? String ?? "",
                                duration: d["duration"] as? Int ?? 5,
                                aspect: d["aspect"] as? String ?? "16:9",
                                resolution: d["resolution"] as? String ?? "standard")
        request.negativePrompt = d["negativePrompt"] as? String ?? ""
        request.fps = d["fps"] as? Int ?? 24
        request.audio = d["audio"] as? Bool ?? true
        request.seed = d["seed"] as? Int ?? -1
        request.turbo = d["turbo"] as? Bool ?? true
        request.imageRef = d["imageRef"] as? String

        var s = AiGenerationSession(generationId: id, request: request)
        s.promptId = d["promptId"] as? String
        s.endpointUsed = d["endpointUsed"] as? String
        s.generationCompleted = d["generationCompleted"] as? Bool ?? false
        s.rewardEarned = d["rewardEarned"] as? Bool ?? false
        s.generationStarted = d["generationStarted"] as? Bool ?? false
        s.adClosedEarly = d["adClosedEarly"] as? Bool ?? false
        s.error = d["error"] as? String
        s.adError = d["adError"] as? String
        s.status = status
        if let path = d["file"] as? String, FileManager.default.fileExists(atPath: path) {
            s.result = URL(fileURLWithPath: path)
        }
        return s
    }
}
